Optimizer state offload to CPU

apaz-cli opened the pull request:
`./scripts/simulate_multi_node_diloco.sh 1 8 src/zeroband/train.py @ configs/10B/H100_devel.toml`

Before:
```py
22:55:18 [DEBUG] [Rank 0] Max memory used: 49695.58 MB
```

After:
```py
22:55:18 [DEBUG] [Rank 0] Max memory used: 49695.58 MB
```

@awgu do you have any idea what's going on here? I added `CPUOffloadPolicy`, but it doesn't seem to be offloading.

awgu commented:
I have been summoned 😮 

I might not be up to date on what the right API is for this, but I would recommend:
1. Adding some assertions to check if your `model.parameters()` are actually on CPU when you expect them to be (e.g. after init)
2. Use memory snapshot which can give you some stack traces of allocations to know why you are still seeing GPU allocation

There might be a public API now, but this is what I usually do:
```
# Add this somewhere early in your init code
torch.cuda.memory._record_memory_history()
...
# Later
snapshot = torch.cuda.memory._snapshot()
with open("snapshot.pickle", "wb") as f:
    pickle.dump(snapshot, f)
# Or something like
snapshot = torch.cuda.memory._snapshot()
with open("snapshot_{torch.distributed.get_rank()}.pickle", "wb") as f:
    pickle.dump(snapshot, f)
```

samsja commented:
thanks  @awgu !!

apaz-cli commented:
@awgu I just spent a couple hours poking at it.

Actually, as far as I can tell, the parameters are not on CPU until the first time `loss.backward()` is called with `model.set_requires_gradient_sync(True)`. Until that point, `param.grad` for all the params is None. I expected it to be there, and I'm sort of wondering what's going on. Do you know what's up with that? Is it documented anywhere?

That's a different question though. The answer to the original question turned out to be that the optimizer actually DID get offloaded. Contrary to my belief. The max memory usage happens right at the beginning when the model is materialized before `fully_shard()`.

So, now I'm trying to figure out how to not actually materialize the full model before sharding.

awgu commented:
I think the unsharded model passed to FSDP is on GPU, which aligns with your observation that the peak memory is before calling `fully_shard`.
https://github.com/PrimeIntellect-ai/prime/blob/adfbb5cc9674c8b2d0baf945b0d186a679a19982/src/zeroband/train.py#L125

At this point:
https://github.com/PrimeIntellect-ai/prime/blob/adfbb5cc9674c8b2d0baf945b0d186a679a19982/src/zeroband/train.py#L175
I would expect all model parameters to be on CPU:
```
for param_name, param in model.named_parameters():
    assert param.device.type == "cpu", f"{param_name} is not on CPU!"
```
Could you check if that is the case?

apaz-cli commented:
@awgu Yep, the parameters are all on CPU.

I now believe the issue was that I uncommented torch.set_default_device(cuda) so the model would load faster. In like 5 seconds, rather than 40. When I comment that out, I see the memory savings.

samsja commented on configs/10B/H100_devel.toml:
```diff
@@ -1,8 +1,8 @@
-name_model = "10B" # "150M"
-# type_model = "llama3"
+name_model = "26B"
+type_model = "llama2"
```
let's create a 26B model config then instead of modifying the 10b_devel

apaz-cli replied:
Fair. This is the one that I'm making changes to, to try to get the model sized to the machine.

samsja commented on src/zeroband/diloco.py:
```diff
@@ -91,6 +91,7 @@ def sync_pseudo_gradient(self, model: nn.Module, fake: bool = False, flag: str =
         global_pg = self.elastic_device_mesh.global_pg
         for i in range(self.config.retry_all_reduce):
             for param_offloaded, param in zip(self.param_list_cpu, model.parameters()):
+                assert isinstance(param_offloaded.grad, DTensor)
```
are all grad always DTensor ? I thought that Lm head and embedding were normal

apaz-cli replied:
They are. I put the assert there largely in case that assumption doesn't hold in the future.

I though that the only thing that was special about he LM head was that after it's gathered, the gradients aren't scattered, they're just immediately used. In which case... I would expect it not to be a dtensor. But it is. Not sure what's going on tbh.

samsja replied:
ah okay cool, then yeah the assert makes sense

samsja commented on src/zeroband/train.py:
```diff
@@ -311,18 +319,17 @@ def train(config: Config):
                     logits = model(tokens=input_ids, block_mask=block_mask).contiguous()
-                    flatten_logits = rearrange(logits, "b seq vocab -> (b seq) vocab")
-                    flatten_labels = rearrange(labels, "b seq -> (b seq)")
+                    flatten_logits = logits.reshape(-1, logits.size(-1))  # b seq vocab -> (b seq) vocab
+                    flatten_labels = labels.reshape(-1)                   # b seq -> (b seq)
```
what is the reason for changing this ?

apaz-cli replied:
We get to drop the einops package

samsja replied:
> We get to drop the einops package.
why ?

> Also theoretically torch compile would be able to work better on it, but we don't compile this.
fair point, actually I think that we are using einops in the attention code, maybe we can makes stuff faster there by removing it

apaz-cli replied:
I believe that this was the only place we used it.

samsja commented on src/zeroband/train.py:
```diff
@@ -342,36 +352,48 @@ def train(config: Config):
             with record_function("Inner allreduce"):
-                dist.all_reduce(tensor=loss_batch, op=dist.ReduceOp.AVG, group=elastic_device_mesh.local_pg)
+                logger.debug("loss allreduce()")
+                # Launch both allreduces at the same time to hide latency
+                loss_allreduce = dist.all_reduce(tensor=loss_batch, op=dist.ReduceOp.AVG, group=elastic_device_mesh.local_pg, async_op=True)
+                if config.optim.z_loss:
+                    z_loss_allreduce = dist.all_reduce(tensor=z_loss_batch, op=dist.ReduceOp.AVG, group=elastic_device_mesh.local_pg, async_op=True)
+
+                assert isinstance(loss_allreduce, torch.distributed.Work)
+                loss_allreduce.wait()
```
I don't think that this optimization matter, this all reduce are extremely fast, its just one number

apaz-cli replied:
I didn't think so either, but figured it would only take a couple seconds.

samsja reviewed (changes requested):
can you break down this pr in two/three PR ? 

one with only the CPU offload code
one with all the others modification. 
one with the lieager kernel updates

Otherwise its hard to review and potentially revert if needed

samsja commented on src/zeroband/data.py:
```diff
@@ -338,7 +337,7 @@ def _load_datasets(
     for _ds in dataset_names.split(","):
         _ds_name, _, _ds_config = _ds.partition(":")
-        _ds_args = {"path": _ds_name}
+        _ds_args: dict[str, Any] = {"path": _ds_name}
```
not sure we need this one tbh

apaz-cli replied:
It seems like a really pointless change, and perhaps it is. I did it because it resolves a type error later on where it's used.

samsja replied:
hmm what type checker are u using ? Not a big fan of adding type for making type checker happy haha

apaz-cli replied:
I'm just using the basic type checking of whatever comes with vscode. Pylance, I believe.

samsja commented on src/zeroband/train.py:
```diff
@@ -144,13 +139,17 @@ def train(config: Config):
         elastic_device_mesh = ElasticDeviceMesh(
-            enable=config.diloco is not None, live_recovery_rank_src=config.ckpt.live_recovery_rank_src
+            enable=config.diloco is not None,
+            live_recovery_rank_src=config.ckpt.live_recovery_rank_src
         )
 
         mp_policy = MixedPrecisionPolicy(
-            param_dtype=torch.bfloat16, reduce_dtype=torch.float32 if config.train.reduce_fp32 else None
+            param_dtype=torch.bfloat16,
+            reduce_dtype=torch.float32 if config.train.reduce_fp32 else None
```
which formatter are u using btw ? Seems that it always do modification on code you did not modify. Maybe we have a different ruff version ?

apaz-cli replied:
I think I just did that one manually. I haven't been using a formatter, just running `ruff check src/` before I put in a PR.

samsja replied:
ah I see, nice, then lets do it in another PR and we are good to go

apaz-cli commented:
Summary of changes:
* Removed einops
* Fixed types
* Added logging to inner training loop
* Changed the arguments on `get_optimizer()`
* Fixed model initialization
* Swap the order of `detach()` and `clone()`
* Overlap loss `all_reduce()`s

samsja commented:
Summary of changes:
Removed einops
Fixed types
Added logging to inner training loop
Changed the arguments on get_optimizer()
Fixed model initialization
Swap the order of detach() and clone()
Overlap loss all_reduce()s

samsja reviewed (approved):
lfgtm
